### PR TITLE
Baked pdf CLI

### PR DESCRIPTION
**Pull request by karenc**
This is the baked pdf CLI branch.  These are the instructions for setting up the CLI (will be updated as the code gets updated):

1. If you're concerned about messing up your work directory, feel free to clone this in a different directory:

   ```
   git clone -b cli-master https://github.com/openstax/cnx-recipes.git cnx-recipes-cli
   cd cnx-recipes-cli
   ```

1. Install docker and docker-compose (docker-compose may already be bundled with docker):

   ```
   ./script/install-docker
   ```

   This should work for people using ubuntu and brew on Mac OS. For everyone else, you probably need to download docker here: https://www.docker.com/products/docker-desktop

   There is no need to run other setup scripts.

1. Remove any outdated docker images (this step only needs to be done once in a while when the code is stable):

   ```
   docker rmi openstax/cnx-recipes:latest openstax/ci-image:latest openstax/princexml:latest openstax/mathify:latest
   ```

   It is ok if this command returns `Error: No such image:`.

   1. Rebuild docker images (this step takes some time):

      ```
      docker-compose build fetch-book neb mathify pdf
      ```

---

These are the instructions for running the CLI (will be updated as the code gets updated):

1. Fetch a book:

   ```
   docker-compose run --rm fetch-book --with-resources intro-business
   ```

   This downloads the cnxml and resources for the book in `./data/intro-business/raw`.

   1. If you want to download from a specific host:

      ```
      docker-compose run --rm -e HOST=katalyst01.cnx.org fetch-book --with-resources intro-business
      ```

1. Generate the single html file for the book:

   ```
   docker-compose run --rm assemble-book intro-business
   ```

     The output is in `./data/intro-business/collection.assembled.xhtml`.

1. Bake the book:

   ```
   docker-compose run --rm bake-book intro-business
   ```

   This creates the baked html file in `./data/intro-business/collection.baked.xhtml`.  It automatically adds the style file `styles/output/<style-name>-pdf.css` to the html file.

   1. If you want to specify your own recipe and / or style css:

      ```
      docker-compose run --rm -e RECIPE=recipes/output/my-recipe.css -e STYLE=styles/output/my-style.css bake-book intro-business
      ``` 

1. (*Optional*) Transform the math in the book:

   ```
   docker-compose run --rm mathify-book intro-business
   ```

   This creates `./data/intro-business/collection.mathified.xhtml`.

1. Create a pdf for the book:

   ```
   docker-compose run --rm build-pdf intro-business
   ```

   This creates `./data/intro-business/collection.pdf`.  It automatically pulls in the style file `styles/output/<style-name>-pdf.css` if it is defined in books.txt.

   1. If you want to pass in a different style css:

      ```
      docker-compose run --rm -e STYLE=recipes/output/my-style.css build-pdf intro-business
      ```

---

I'll rebase and fix up the commits when people have tried out the CLI.

### CODE REVIEW

**Comment by karenc** [comment]
Could someone add some Katalyst devs as reviewers?  I'm not sure who's relevant.

File: script/bootstrap [hunk 1]
```diff
@@ -105,71 +173,20 @@ if [[ ! ${BOOTSTRAP_ALREADY_RAN} ]]; then
     }
   fi
 
+  # Install docker
+  if [[ -z "$(command -v docker)" ]]; then
+    if [[ -n "$(command -v apt-get)" ]]; then
+      do_progress "Installing docker" \
+        sudo apt-get install docker.io
+    fi
```
Comment (philschatz) [inline script/bootstrap]:
```suggestion
    else
```

**Comment by brittany-johnson** [comment]
First time running `docker-compose run --rm fetch-html --with-resources` did take a while like you expected Karen but it worked fine and was faster the second time. 
Fetching with and without resources on different servers also worked well, I really like that we can specify that when running the command. I also really like the countdown when fetching and the progress bar. 
Overall, all of the commands above worked well and it will be very easy to transition to the CLI. 

The only question I have is that I already had docker install, but I ran `./scipt/install-docker` anyway, and I didn't get any output. If someone already has docker installed, I'm not sure if it would be useful to have a message output letting them know that it's already installed? What do you think?

**Comment by tomjw64** [comment]
Same experience as BrittanyJ with the time to run. Slow first run, nice speed on the following. If i'm downloading lots of books, i would like the option to skip the `File '/out/intro-business/collection.assembled.xhtml' already exists. Would you like to replace it? [y/N]:` prompt. In most cases, I know what I'm doing when I fetch the book.
One frustrating thing I ran into was that the fetch-html docker script makes `root` both the user and group for the downloaded files. It might be nicer to make the group `docker` or `$(whoami)` so that those of us that give non-root users docker permissions can use it a bit easier.
Another question I have is: Do we want the scripts to still be runnable without the context of a docker container. A standalone `./script/fetch-html intro-business` now fails with the following traceback:
```
Traceback (most recent call last):
  File "/usr/local/bin/neb", line 11, in <module>
    load_entry_point('nebuchadnezzar', 'console_scripts', 'neb')()
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/app/nebu/cli/_common.py", line 54, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/app/nebu/cli/get.py", line 119, in get
    os.mkdir(str(output_dir))
FileNotFoundError: [Errno 2] No such file or directory: '/app/data/intro-business/raw'
./script/fetch-html: ERROR: could not run [docker-compose run --rm neb neb get -d ./data/intro-business/raw easyvm5.cnx.org col25734 ]
```

**Comment by karenc** [comment]
> First time running `docker-compose run --rm fetch-html --with-resources` did take a while like you expected Karen but it worked fine and was faster the second time.
> Fetching with and without resources on different servers also worked well, I really like that we can specify that when running the command. I also really like the countdown when fetching and the progress bar.
> Overall, all of the commands above worked well and it will be very easy to transition to the CLI.
> 
> The only question I have is that I already had docker install, but I ran `./scipt/install-docker` anyway, and I didn't get any output. If someone already has docker installed, I'm not sure if it would be useful to have a message output letting them know that it's already installed? What do you think?

@brittany-johnson Thanks for taking the time to try this out.  You're right about the `install-docker` script, I should output a message saying "Docker is already installed."  I'll change that.

I'm supposed to write a wrapper script that calls "docker-compose run --rm" etc, do you have any suggestions about what it should be called?

e.g.

```
./script/<name> fetch-html ...
./script/<name> bake-book ...
./script/<name> mathify-book ...
./script/<name> build-pdf ...
```

**Comment by helenemccarron** [comment]
on a fresh cloned version of recipes, I get this: 
``` 
helene@McBook > docker-compose build fetch-html neb mathify pdf                                
ERROR: No such service: fetch-html
```
`docker-compose run --rm fetch-book --with-resources intro-business` did work though.

I ran`docker-compose run --rm bake-book intro-business` and it worked:
```
==> Adding style file ./styles/output/intro-business-pdf.css
==> Validating ./data/intro-business/collection.baked.xhtml
```
then ran 
 `docker-compose run --rm mathify-book intro-business` which worked: 



then ` docker-compose run --rm build-pdf intro-business`

which worked but "lost" the styles:

**Comment by karenc** [comment]
> Same experience as BrittanyJ with the time to run. Slow first run, nice speed on the following. If I'm downloading lots of books, I would like the option to skip the `File '/out/intro-business/collection.assembled.xhtml' already exists. Would you like to replace it? [y/N]:` prompt. In most cases, I know what I'm doing when I fetch the book.

@tomjw64 Thanks for your feedback!  Yeah, I think in the script, I can remove `collection.assembled.xhtml` so it wouldn't prompt the user.

> One frustrating thing I ran into was that the fetch-html docker script makes `root` both the user and group for the downloaded files. It might be nicer to make the group `docker` or `$(whoami)` so that those of us that give non-root users docker permissions can use it a bit easier.

This is because inside the container, the scripts run as root so any files they create also belong to root.  I'll fix this.

> Another question I have is: Do we want the scripts to still be runnable without the context of a docker container. A standalone `./script/fetch-html intro-business` now fails with the following traceback:
> 
> ```
> Traceback (most recent call last):
>   File "/usr/local/bin/neb", line 11, in <module>
>     load_entry_point('nebuchadnezzar', 'console_scripts', 'neb')()
>   File "/usr/local/lib/python3.6/site-packages/click/core.py", line 764, in __call__
>     return self.main(*args, **kwargs)
>   File "/usr/local/lib/python3.6/site-packages/click/core.py", line 717, in main
>     rv = self.invoke(ctx)
>   File "/usr/local/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>   File "/usr/local/lib/python3.6/site-packages/click/core.py", line 956, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/usr/local/lib/python3.6/site-packages/click/core.py", line 555, in invoke
>     return callback(*args, **kwargs)
>   File "/usr/local/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func
>     return f(get_current_context(), *args, **kwargs)
>   File "/app/nebu/cli/_common.py", line 54, in wrapper
>     return func(*args, **kwargs)
>   File "/usr/local/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func
>     return f(get_current_context(), *args, **kwargs)
>   File "/app/nebu/cli/get.py", line 119, in get
>     os.mkdir(str(output_dir))
> FileNotFoundError: [Errno 2] No such file or directory: '/app/data/intro-business/raw'
> ./script/fetch-html: ERROR: could not run [docker-compose run --rm neb neb get -d ./data/intro-business/raw easyvm5.cnx.org col25734 ]
> ```

I did think a bit about how to not make the scripts fail when you're not running it in docker, but I didn't test it.  Is this a "must work" case or best effort?  It would be hard to make sure the scripts work locally on different people's machines.

Do you expect `./script/fetch-book` to work locally with `docker-compose run` inside the script or do you expect the script to not use `docker` at all?

**Comment by Adrianltd** [comment]
@karenc 
1. When I run `docker rmi openstax/cnx-recipes:latest openstax/ci-image:latest openstax/princexml:latest openstax/mathify:latest` I've received this:
>
>Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Delete http://%2Fvar%2Frun%2Fdocker.sock/v1.39/images/openstax/cnx-recipes:latest: dial unix /var/run/docker.sock: connect: permission denied
> Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Delete http://%2Fvar%2Frun%2Fdocker.sock/v1.39/images/openstax/ci-image:latest: dial unix /var/run/docker.sock: connect: permission denied
> Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Delete http://%2Fvar%2Frun%2Fdocker.sock/v1.39/images/openstax/princexml:latest: dial unix /var/run/docker.sock: connect: permission denied
> Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Delete http://%2Fvar%2Frun%2Fdocker.sock/v1.39/images/openstax/mathify:latest: dial unix /var/run/docker.sock: connect: permission denied
>
and  when running`docker-compose build fetch-book neb mathify pdf` I've got this:
>
> Building fetch-book
> ERROR: Couldn't connect to Docker daemon at http+docker://localunixsocket - is it running?
>
> If it's at a non-standard location, specify the URL with the DOCKER_HOST environment variable.
>
I had to run those two commands with `sudo`, no errors then. Other scripts have to be run with `sudo` too, but they output errors.

This `docker-compose run --rm fetch-book --with-resources intro-business` says:
>
> WARNING: The PWD variable is not set. Defaulting to a blank string.
> ==> Fetching intro-business from easyvm5.cnx.org
> Error: unknown environment 'easyvm5.cnx.org'
> /code/script/fetch-book: ERROR: could not run [docker-compose run --rm neb neb get -r -d /out/intro-business/raw easyvm5.cnx.org col25734 ]
>

After trying `docker-compose run --rm neb neb get -r -d /out/intro-business/raw easyvm5.cnx.org col25734` i get:
>
> WARNING: The PWD variable is not set. Defaulting to a blank string.
> Usage: neb get [OPTIONS] ENV COL_ID COL_VERSION
> Try "neb get --help" for help.
> 
> Error: Missing argument "COL_VERSION".
>

This `docker-compose run --rm assemble-book intro-business` says:
>
> WARNING: The PWD variable is not set. Defaulting to a blank string.
> ==> Creating HTML file for intro-business
> Usage: neb assemble [OPTIONS] INPUT_DIR OUTPUT_DIR
> Try "neb assemble --help" for help.
> 
> Error: Invalid value for "INPUT_DIR": Path "/out/intro-business/raw" does not exist.
> /code/script/assemble-book: ERROR: could not run [docker-compose run --rm neb neb assemble /out/intro-business/raw /out/intro-business]
> 

This `docker-compose run --rm fetch-book /bin/bash -c 'cat config/neb-config'` runs forever with no output, same with using `sudo`

**Comment by wilkus27** [comment]
@karenc I'm pasting here part of our conversation from Slack to be documented also on GitHub:

image
image
image (1)
image
image (2)
image
image (3)
image
image (4)
image
image (5)
image

Fixed with last commits :)

**Comment by karenc** [comment]
@Adrianltd 

> 1. When I run `docker rmi openstax/cnx-recipes:latest openstax/ci-image:latest openstax/princexml:latest openstax/mathify:latest` I've received this:
> 
> > Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Delete http://%2Fvar%2Frun%2Fdocker.sock/v1.39/images/openstax/cnx-recipes:latest: dial unix /var/run/docker.sock: connect: permission denied
> > Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Delete http://%2Fvar%2Frun%2Fdocker.sock/v1.39/images/openstax/ci-image:latest: dial unix /var/run/docker.sock: connect: permission denied
> > Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Delete http://%2Fvar%2Frun%2Fdocker.sock/v1.39/images/openstax/princexml:latest: dial unix /var/run/docker.sock: connect: permission denied
> > Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Delete http://%2Fvar%2Frun%2Fdocker.sock/v1.39/images/openstax/mathify:latest: dial unix /var/run/docker.sock: connect: permission denied

I updated the message in `./script/install-docker` to say "You need to logout and log back in for it to work.".  The reason for this is your user doesn't automatically gets permission to run docker.  Your user needs to be added to the `docker` group (which the script does) and the system doesn't see the new group unless you login again.

> 
> and when running`docker-compose build fetch-book neb mathify pdf` I've got this:
> 
> > Building fetch-book
> > ERROR: Couldn't connect to Docker daemon at http+docker://localunixsocket - is it running?
> > If it's at a non-standard location, specify the URL with the DOCKER_HOST environment variable.
> 
> I had to run those two commands with `sudo`, no errors then. Other scripts have to be run with `sudo` too, but they output errors.

As explained above, if you logout and login again, you shouldn't have to run the commands with `sudo`.

> 
> This `docker-compose run --rm fetch-book --with-resources intro-business` says:
> 
> > WARNING: The PWD variable is not set. Defaulting to a blank string.
> > ==> Fetching intro-business from easyvm5.cnx.org
> > Error: unknown environment 'easyvm5.cnx.org'
> > /code/script/fetch-book: ERROR: could not run [docker-compose run --rm neb neb get -r -d /out/intro-business/raw easyvm5.cnx.org col25734 ]

I think @wilkus27 has seen this (unknown environment 'easyvm5.cnx.org') before.  I asked her to pull the latest changes and rebuild some of the images and it worked for her IIRC:

```
docker-compose build fetch-book neb
```

I haven't seen "The PWD variable is not set" warning before.  What environment are you running?  Ubuntu?  If you do `echo $PWD` in a terminal, is it empty?

> 
> After trying `docker-compose run --rm neb neb get -r -d /out/intro-business/raw easyvm5.cnx.org col25734` i get:
> 
> > WARNING: The PWD variable is not set. Defaulting to a blank string.
> > Usage: neb get [OPTIONS] ENV COL_ID COL_VERSION
> > Try "neb get --help" for help.
> > Error: Missing argument "COL_VERSION".
> 

If possible we shouldn't be running neb directly, but to explain what went wrong here, it didn't print out an empty string at the end, so it should be:

```
docker-compose run --rm neb neb get -r -d /out/intro-business/raw easyvm5.cnx.org col25734 ''
```

> This `docker-compose run --rm assemble-book intro-business` says:
> 
> > WARNING: The PWD variable is not set. Defaulting to a blank string.
> > ==> Creating HTML file for intro-business
> > Usage: neb assemble [OPTIONS] INPUT_DIR OUTPUT_DIR
> > Try "neb assemble --help" for help.
> > Error: Invalid value for "INPUT_DIR": Path "/out/intro-business/raw" does not exist.
> > /code/script/assemble-book: ERROR: could not run [docker-compose run --rm neb neb assemble /out/intro-business/raw /out/intro-business]
> 
> This `docker-compose run --rm fetch-book /bin/bash -c 'cat config/neb-config'` runs forever with no output, same with using `sudo`

This probably has to do with "the PWD variable is not set".  If you let me know what kind of environment you're using to run docker-compose, I can try to fix this...

**Comment by Adrianltd** [comment]
@karenc 

> I updated the message in ./script/install-docker to say "You need to logout and log back in for it to work.". The reason for this is your user doesn't automatically gets permission to run docker. Your user needs to be added to the docker group (which the script does) and the system doesn't see the new group unless you login again.
I've just cloned fresh repo again, did restart the PC, still have to use `sudo`.

> I think @wilkus27 has seen this (unknown environment 'easyvm5.cnx.org') before. I asked her to pull the latest changes and rebuild some of the images and it worked for her IIRC:
> 
> ```
> docker-compose build fetch-book neb
> ```

Done that, still the same.

> I haven't seen "The PWD variable is not set" warning before. What environment are you running? Ubuntu? If you do
> 
> ```
> printenv | grep PWD
> ```

This outputs:
>
>PWD=/home/adriank/cnx-recipes-cli
>OLDPWD=/home/adriank
>
Running `docker-compose run --rm neb neb get -r -d /out/intro-business/raw easyvm5.cnx.org col25734 ''` 
outputs this:
>
>`WARNING: The PWD variable is not set. Defaulting to a blank string.
>Error: content unavailable for 'col25734/'
>

I'm running `Ubuntu 18.04.1 LTS`.
Thank you for your time fixing all of this, I'm not sure why I'm having those issues. It's freshly cloned repo again, all scripts are run step by step as docs say. Natalia, as far as I know, is running the same environment as I do, but what she's done is not helping in my case.

**Comment by karenc** [comment]
@Adrianltd

> 
> > I updated the message in ./script/install-docker to say "You need to logout and log back in for it to work.". The reason for this is your user doesn't automatically gets permission to run docker. Your user needs to be added to the docker group (which the script does) and the system doesn't see the new group unless you login again.
> > I've just cloned fresh repo again, did restart the PC, still have to use `sudo`.

There are a few things I want to check regarding running docker without sudo:

1. When you do `groups`, do you see `docker`?

1. Are you able to run `docker images` without `sudo`?

1. Are there files in cnx-recipes owned by root?  `find <path-to-cnx-recipes> -user root`

   1. If you change all the files owned by root to your user:

      ```
      find <path-to-cnx-recipes> -user root -exec chown $USER '{}' \;
      ```

      Are you able to run the `docker-compose` commands without `sudo`?

> 
> > I think @wilkus27 has seen this (unknown environment 'easyvm5.cnx.org') before. I asked her to pull the latest changes and rebuild some of the images and it worked for her IIRC:
> > ```
> > docker-compose build fetch-book neb
> > ```
> 
> Done that, still the same.

I think this is definitely related to "The PWD variable is not set".
I searched around for this message and found that it might be a specific way `docker-compose` is installed.

If you do `head $(which docker-compose)`, what do you see?

Mine looks like:

```
#!/usr/bin/python
# EASY-INSTALL-ENTRY-SCRIPT: 'docker-compose==1.17.1','console_scripts','docker-compose'
__requires__ = 'docker-compose==1.17.1'
import re
import sys
from pkg_resources import load_entry_point

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(
```

If your `docker-compose` looks different, do you remember how you installed
`docker-compose`?

> 
> > I haven't seen "The PWD variable is not set" warning before. What environment are you running? Ubuntu? If you do
> > ```
> > printenv | grep PWD
> > ```
> 
> This outputs:
> 
> > PWD=/home/adriank/cnx-recipes-cli
> > OLDPWD=/home/adriank

That looks good to me.

> 
> Running `docker-compose run --rm neb neb get -r -d /out/intro-business/raw easyvm5.cnx.org col25734 ''`
> outputs this:
> 
> > `WARNING: The PWD variable is not set. Defaulting to a blank string.
> > Error: content unavailable for 'col25734/'
> 
> I'm running `Ubuntu 18.04.1 LTS`.
> Thank you for your time fixing all of this, I'm not sure why I'm having those issues. It's freshly cloned repo again, all scripts are run step by step as docs say. Natalia, as far as I know, is running the same environment as I do, but what she's done is not helping in my case.

Thanks for working with me on this!  I was hoping that using docker would solve
a lot of the installation problems, but it may not be that simple.

**Comment by karenc** [comment]
I forgot to announce here.  I've fixed the output file owned by root problem.  So new files created should be owned by the user.  https://github.com/openstax/cnx-recipes/pull/1039/commits/0eb748a0608ad0e42b9fa3856127e8b2b5bda23b

**Comment by Adrianltd** [comment]
> 1. When you do `groups`, do you see `docker`?
No :O 

> 2. Are you able to run `docker images` without `sudo`?
No

> 3. Are there files in cnx-recipes owned by root?  `find <path-to-cnx-recipes> -user root`
There's this folder only `cnx-recipes-cli/data/intro-business`

> 1. If you change all the files owned by root to your user:
>    ```
>    find <path-to-cnx-recipes> -user root -exec chown $USER '{}' \;
>    ```
>    Are you able to run the `docker-compose` commands without `sudo`?

No, after running `docker-compose build fetch-book neb mathify pdf`:
> Building fetch-book
> ERROR: Couldn't connect to Docker daemon at http+docker://localunixsocket - is it running?
>
> If it's at a non-standard location, specify the URL with the DOCKER_HOST environment variable.

> If you do `head $(which docker-compose)`, what do you see?
> 
> Mine looks like:

Mine is the same.

So for some reason I do not have docker in groups and this seems to be the case.
I'll try to uninstall all the Docker files and images I have once again, maybe I've missed something, and then I'll run the setup again.

**Comment by karenc** [comment]
@Adrianltd 

> So for some reason I do not have docker in groups and this seems to be the case.
> I'll try to uninstall all the Docker files and images I have once again, maybe I've missed something, and then I'll run the setup again.

In the `install-docker` script, I tried to add the user to the `docker` group in the script by doing:

```
sudo usermod -a -G docker "$USER"
```

^ But this only works if you run `install-docker` **without** sudo and if you restart your session.

(I'm going to fix up the script so it works even when run with sudo.)

**Comment by Adrianltd** [comment]
@karenc As said, I have successfully removed all of the Docker files and reinstalled it again.
I've used `dpkg -l | grep -i docker` to find whatever docker files I have.
Then I've used `sudo apt-get purge -y dockerFileNames` and `sudo apt-get autoremove -y dockerFileNames`
Then I have removed all images, conf files etc.
`sudo rm -rf /var/lib/docker`
`sudo rm -rf /var/run/docker.sock`
`sudo rm -rf /etc/docker`
And reinstalled the docker `./script/install-docker`. And that worked. Almost :P

When I run `docker-compose build fetch-book neb mathify pdf` everything works excluding:
> Building mathify
> ERROR: error fetching: fatal: Couldn't find remote ref change-svg-font
>: exit status 128

So I cannot run `docker-compose run --rm mathify-book intro-business` I have this:
> ==> Converting math in /out/intro-business/collection.baked.xhtml to svg
> Building mathify
> ERROR: error fetching: fatal: Couldn't find remote ref change-svg-font
> : exit status 128
> /code/script/mathify-book: ERROR: could not run [docker-compose run --rm mathify /bin/bash -c node --max_old_space_size=8192 typeset/start -i "/out/intro-business/collection.baked.xhtml" -o "/out/intro-business/collection.mathified.xhtml" -f "svg" && cp -r node_modules/mathjax "/out/intro-business" && chmod a+w "/out/intro-business/collection.mathified.xhtml" && chown -R 1000:1000 "/out/intro-business"]

All of the other scripts work perfectly now ;)

**Comment by Adrianltd** [comment]
@karenc update: I have updated my branch with your latest changes (those made on 16.05) and those scripts above work! No more issues, uff!

**Comment by karenc** [comment]
> @karenc update: I have updated my branch with your latest changes (those made on 16.05) and those scripts above work! No more issues, uff!

@Adrianltd That's great news!!  Thanks for your patience while testing this...! :sweat_smile:  Just to give you a heads up, there are some `mathify-book` problems that we're still looking at.

**Comment by karenc** [comment]
The baked pdf CLI has been given the go-ahead by the books team.  I'm just going to clean up the commits and when it's ready, change it to ready for review (not work in progress).

**Comment by karenc** [comment]
Successfully cleaned up 63 commits and made 12 meaningful commits.  Removed unnecessary changes and combined some of them.  There are changes that aren't directly related to the baked pdf CLI, so it's best to not squash.

**Pull request merged**